layer1/executor/task_executor.go

```diff
@@ -0,0 +1,354 @@
+	TxsBackup map[string]*types.Transaction `json:"transactions_backup"`
+	txWatcher transaction.Watcher           `json:"-"`
+	database  *db.Database                  `json:"-"`
+	logger    *logrus.Entry                 `json:"-"`
+}
+
+// newTaskExecutor creates a new TaskExecutor instance and recover the previous state from DB.
+func newTaskExecutor(txWatcher transaction.Watcher, database *db.Database, logger *logrus.Entry) (*TaskExecutor, error) {
+	taskExecutor := &TaskExecutor{
+		TxsBackup: make(map[string]*types.Transaction),
+		txWatcher: txWatcher,
+		database:  database,
+		logger:    logger,
+	}
+
+	err := taskExecutor.loadState()
+	if err != nil {
+		taskExecutor.logger.Warnf("could not find previous State: %v", err)
+		if err != badger.ErrKeyNotFound {
+			return nil, err
+		}
+	}
+
+	return taskExecutor, nil
+}
+
+// addTxBackup adds txn to the backup map for recovery.
+func (te *TaskExecutor) addTxBackup(uuid string, tx *types.Transaction) error {
+	te.Lock()
+	defer te.Unlock()
+	te.TxsBackup[uuid] = tx
+	return te.persistState()
+}
+
+// removeTxBackup removes txn from backup map.
+func (te *TaskExecutor) removeTxBackup(uuid string) error {
+	te.Lock()
+	defer te.Unlock()
+	delete(te.TxsBackup, uuid)
+	return te.persistState()
+}
+
+// getTxBackup retrieves txn from backup map.
+func (te *TaskExecutor) getTxBackup(uuid string) (*types.Transaction, bool) {
+	te.RLock()
+	defer te.RUnlock()
+	txn, present := te.TxsBackup[uuid]
+	return txn, present
+}
+
+// handleTaskExecution It is basically an abstraction to handle the task execution in a separate process.
+func (te *TaskExecutor) handleTaskExecution(mainCtx context.Context, task tasks.Task, name string, taskId string, start uint64, end uint64, allowMultiExecution bool, subscribeOptions *transaction.SubscribeOptions, database *db.Database, logger *logrus.Entry, eth layer1.Client, contracts layer1.AllSmartContracts, taskResponseChan tasks.InternalTaskResponseChan) {
+	defer task.Close()
+	err := te.processTask(mainCtx, task, name, taskId, start, end, allowMultiExecution, subscribeOptions, database, logger, eth, contracts, taskResponseChan)
+	// Clean up in case the task was killed
+	if task.WasKilled() {
+		task.GetLogger().Trace("task was externally killed, removing tx backup")
+		te.removeTxBackup(task.GetId())
+	}
+	task.Finish(err)
+}
+
+// processTask processes all the stages of the task execution.
+func (te *TaskExecutor) processTask(mainCtx context.Context, task tasks.Task, name string, taskId string, start uint64, end uint64, allowMultiExecution bool, subscribeOptions *transaction.SubscribeOptions, database *db.Database, logger *logrus.Entry, eth layer1.Client, contracts layer1.AllSmartContracts, taskResponseChan tasks.InternalTaskResponseChan) error {
+	taskCtx, cf := context.WithCancel(mainCtx)
+	defer cf()
+
+	err := task.Initialize(taskCtx, cf, database, logger, eth, contracts, name, taskId, start, end, allowMultiExecution, subscribeOptions, taskResponseChan)
```

Comment:
Cover err case in tests if possible - should unwind the main.

Reply:
Will cover in tests. The errors that we can have during the task execution are not supposed to unwind the main. The reasoning behind that is we don't need to kill the validator if a task was unsuccessfully executed. 
Super silly question: by `should unwind the main` do you mean kill the validator?